Fix multiple usability issues in the texture region editor

YuriSizov opened the pull request:
Fixes https://github.com/godotengine/godot/issues/75291. Fixes https://github.com/godotengine/godot/issues/80403 (a duplicate, actually). Fixes https://github.com/godotengine/godot/issues/75989 (at least, I can't reproduce it with the PR applied).

Changes include:

- Correctly display atlas textures when they are used by other supported objects (like nodes or styleboxes). We used to draw everything through a `CanvasTexture`, but it can't handle custom drawing logic of textures that it holds, such as `AtlasTexture`. So I reworked it to avoid relying on the `CanvasTexture`.

- Make region handles easier to interact with when editing NinePatchRect and StyleBoxTexture, which also have bars to adjust margins. We used to check handles last, which means bars would steal clicks. Since bars are normally long and can be clicked in plenty of places, and handles are small and located in precise positions, I switched it so handles are checked first.

- While testing I've noticed that a few things weren't properly initialized or restored. To name a couple, the default snap mode from the editor settings wasn't correctly applied, and the panning offset wouldn't completely update when opening the dialog.

- Various code quality changes. Moved all property initializations into the header. Removed unused methods, changed access modifiers of some remaining. Renamed everything. Extracted a few bits of logic into dedicated methods and unified how those are used. And other such changes.

- Add a check to `AtlasTexture::get_image()` to avoid bliting with invalid size values. It spammed a lot in the editor when using the region editor otherwise.

Apparently, you can use Godot's nose to create a nice ninepatch:

<img width="650" alt="NVIDIA_Share_2023-08-08_23-07-11" src="https://github.com/godotengine/godot/assets/11782833/ad8460c3-658a-4e52-9fdc-e35262a6b6f5">

akien-mga reviewed (approved):
Code style looks good.

It's a ton of changes, I didn't review the code in depth.

YuriSizov commented on editor/plugins/texture_region_editor_plugin.cpp:
```diff
@@ -38,23 +38,24 @@
 #include "scene/resources/atlas_texture.h"
 
-void draw_margin_line(Control *edit_draw, Vector2 from, Vector2 to) {
+void draw_margin_line(Control *texture_overlay, Vector2 from, Vector2 to) {
```
I think this was done on purpose (with the original name). This is a class-less method that only exists in the cpp file, and in this case `edit_draw`/`texture_overlay` actually refer to the class member which is passed to this method. For consistency I'll rename the parameters, but I'd like some policy regarding using such helper methods, is it even useful?

YuriSizov committed:
Fix multiple usability issues in the texture region editor

- Correctly display atlas textures when used by other objects.
- Make region handles easier to hit in ninepatchable objects.
- Correctly initialize and restore various visual properties.
- Improve code quality.

adamscott commented on editor/plugins/texture_region_editor_plugin.cpp:
```diff
@@ -38,64 +38,67 @@
-			from,
-			to,
+	p_texture_overlay->draw_line(
+			p_from,
+			p_to,
 			EditorNode::get_singleton()->get_theme_base()->get_theme_color(SNAME("mono_color"), SNAME("Editor")).inverted() * Color(1, 1, 1, 0.5),
 			Math::round(2 * EDSCALE));
 
-	while (from.distance_squared_to(to) > 200) {
-		edit_draw->draw_line(
-				from,
-				from + line,
+	while (p_from.distance_squared_to(p_to) > 200) {
+		p_texture_overlay->draw_line(
+				p_from,
+				p_from + line,
 				EditorNode::get_singleton()->get_theme_base()->get_theme_color(SNAME("mono_color"), SNAME("Editor")),
 				Math::round(2 * EDSCALE));
 
-		from += line * 2;
+		p_from += line * 2;
 	}
 }
```
There's a lot of magic numbers used. I know that it was in the previous code, but could we use some `constexpr` to express those numbers?

YuriSizov replied:
I can do another code quality pass as a follow up. There are more things that can be improved here, but I'd rather not dilute this PR too much — it's pretty messy as it is (I should've acted with more restraint to begin with).

adamscott replied:
Maybe just for that function. It's pretty egregious. Like, what does the 200 mean on line 7? And the `* 2` afterwards.

YuriSizov replied:
The `*2` is because it's a dashed line. We first draw a solid line then draw dashes on top of it with a contrasting color. `line` is the size of the dash, so we draw one unit of `line` and skip another unit of `line`.

As for 200, I'm not completely sure myself.

YuriSizov replied:
I've done a little cleanup and added comments to explain the logic.

adamscott reviewed (approved):
As @akien-mga said, this seems fine. Didn't review the code extensively, but it looks good.

kleonc commented on editor/plugins/texture_region_editor_plugin.h:
```diff
@@ -66,45 +67,47 @@ class TextureRegionEditor : public AcceptDialog {
 	bool updating_scroll = false;
 
-	int snap_mode = 0;
+	int snap_mode = SnapMode::SNAP_NONE;
```
Couldn't it be:
```suggestion
	SnapMode snap_mode = SNAP_NONE;
```
?

kleonc reviewed (commented):
Skimmed through the code, definitely looks like an improvement! :+1:

YuriSizov commented on scene/resources/atlas_texture.cpp:
```diff
@@ -248,6 +248,9 @@ Ref<Image> AtlasTexture::get_image() const {
 	if (!atlas.is_valid() || !atlas->get_image().is_valid()) {
 		return Ref<Image>();
 	}
+	if (region.size.x <= 0 || region.size.y <= 0) {
+		return Ref<Image>();
+	}
 
 	return atlas->get_image()->get_region(region);
```
I agree with the suggestion and reasoning for it, made the changes.

For the second half, I don't know how it supposed to work, but I think that for the purposes of the region editor we don't want to have the margin applied. I think it's supposed to be added after the region. But yeah, it's not new to this PR, so I'd leave it for another day if someone has a better idea.

kleonc replied:
> For the second half, I don't know how it supposed to work, but I think that for the purposes of the region editor we don't want to have the margin applied. I think it's supposed to be added after the region. But yeah, it's not new to this PR, so I'd leave it for another day if someone has a better idea.

Yeah, I agree. Margin shouldn't be shown in the region selector, it's indeed "added after region". What would make sense for the margin is e.g. to have a separate "margin editor" dialog (openable by "Edit Margin" button in the inspector) where the already selected region would be shown and it would allow to only drag the margins on top of the region.
But yeah, that's for a different PR.

kleonc reviewed (approved):
I've done some testing, LGTM!
Tried multi-nested AtlasTextures with non-zero margins. Even for these both texture rendering and region selecting is working correctly.
I've also played a little with StyleBoxTexture or NinePatchRect, haven't had any issues.

Note there are still cases where the texture (for which the region is being edited) will be rendered "incorrectly". E.g. when selecting a region for a CanvasTexture with an AtlasTexture assigned to its diffuse texture, such texture would't be rendered "correctly" within the region editor. But such specific combination doesn't work properly even outside of the region editor (I think using AtlasTextures within CanvasTexture is not supported), meaning that's not an issue with the region editor itself. 

AFAICT, what's supported seems to work just fine. :+1:
(haven't tested everything though; but for sure things work better than before :upside_down_face:)

---

There's this small issue when editing a `(0, 0, 0, 0)` region:
![godot windows editor x86_64_QwJWiFQR4i](https://github.com/godotengine/godot/assets/9283098/0beb9a15-4553-43e0-8dcd-674e46aa6691)
After opening the region editor it shows as if the whole texture was selected as the region:
![wgBMFXaxMP](https://github.com/godotengine/godot/assets/9283098/5b176ffa-610f-45b2-8f1f-359ea86e58b3)
Closing the region editor without changes makes the region remain being `(0, 0, 0, 0)`. This might be surprising given within the texture editor the whole texture was shown to be selected.
Any slight region-drag within the region editor is needed to make the region be set accordingly within the Sprite2D etc. (even when snapping so no effective change):
![LMEDXVfCLD](https://github.com/godotengine/godot/assets/9283098/4d242d17-b0cf-467c-8c98-813460909e8f)
Not sure if anything/what needs to be done about this.

YuriSizov commented:
> Not sure if anything/what needs to be done about this.

I can revert some changes, because that's on me. I unified some of that logic that used to treat these as special cases (or rather other empty regions as special cases). I think, though, that it's more intuitive for an empty region to equal the whole image. At least it is more user friendly when opening the region editor, in my mind.

But I guess there is a different between an empty region and and empty region with a "use region" toggle. So I dunno. We can keep it as is and see if there are complaints after the change, I guess.

akien-mga commented:
Thanks!

ramireshimself commented:
Help. please"
Is there a way I can dock the Region Editor?
![image](https://github.com/godotengine/godot/assets/125474408/2d7938f6-bc75-4bed-83ac-94f83f57aedd)
![image](https://github.com/godotengine/godot/assets/125474408/ce60a076-6cbe-4db9-b3e0-6bc911b713b9)
I´m foloing a tutorial and its reali anoying that I can't dock the editor as it is in the video and I have to open and close it every time I change the timeline for each region edit.

YuriSizov commented:
@ramireshimself You cannot dock it, no. It used to be a contextual editor on the bottom panel, but this was changed in Godot 4.

ramireshimself commented:
Thanks for the awenser. So there is no simmilar way to do like before? I have to do it one by one, closing, keyfaming and opening again, and edit, etc, etc?

KoBeWi commented:
You can input the value in the inspector.

ramireshimself commented:
thnks, I´ll try.
